### PR TITLE
Fixes #134 regex routes

### LINKED ISSUE #134: Path cmdlets need ParameterSet for RegEx Paths
The way the code works at the moment you can pass a straight regular expression as the path and the app will use that but it probably should be split into another ParameterSet that takes a RegEx as the Path parameter to allow for better documentation and clarity.

Related: RFC 001 (#120)

### PR DESCRIPTION
**Pull request by Tiberriver256**
## WIP Checklist

- [x] Fully implement RFC 001
- [x] Add about_Routing.md

## Pull Request Description

Finally got around to finishing the full implementation of what was discussed in RFC 001 (#120). You can now pass in an actual [RegEx] type as your path if you would like. Otherwise it will still support regular strings and string patterns.

I'm going to write up the docs for this over the next couple of days but wanted to open this up as WIP to get some feedback on the changes I made to the Polaris class public property **ScriptBlockRoutes** which are:

1. I renamed it from ScriptBlockRoutes to **Routes**
2. I changed it from a key value of key values to be a new class called [PolarisRoute]

The [PolarisRoute] class has several niceties that I felt were missing in the old system and I believe I have simplified things quite a bit. Instead of having to loop through keys and convert path strings to regular expressions at run time we convert the path to regular expression in the setter function of the path property on the [PolarisRoute] which 1) makes for easier debugging as you can see what the URL is getting matched against for your given path 2) saves some processing time as the paths are only converted to regular expressions once.

Example case given the following routes:

```pwsh
New-PolarisRoute -Path "/users/:userId/books/:bookId" -Method "GET" -Scriptblock {
    $Response.Send(($Request.Parameters | ConvertTo-JSON))
}

New-PolarisRoute -Path "/flights/:from-:to" -Method "GET" -Scriptblock {
    $Response.Send(($Request.Parameters | ConvertTo-JSON))
}

New-PolarisRoute -Path "/flights/:from.:to" -Method "GET" -Scriptblock {
    $Response.Send(($Request.Parameters | ConvertTo-JSON))
}

New-PolarisRoute -Path "/user/(?<userId>\d+)" -Method "GET" -Scriptblock {
    $Response.Send(($Request.Parameters | ConvertTo-JSON))
}

New-PolarisRoute -Path ([Regex]::new("/userRegex/(?<userId>\d+)")) -Method "GET" -Scriptblock {
    $Response.Send(($Request.Parameters | ConvertTo-JSON))
}

New-PolarisRoute -Path "/" -Method "GET" -Scriptblock {
    $Response.Send("root")
}

New-PolarisRoute -Path "/about" -Method "GET" -Scriptblock {
    $Response.Send("about")
}

New-PolarisRoute -Path "/random.text" -Method "GET" -Scriptblock {
    $Response.Send("random.text")
}

New-PolarisRoute -Path "/12?34" -Method "GET" -Scriptblock {
    $Response.Send("12?34")
}

New-PolarisRoute -Path "/ab+cd" -Method "GET" -Scriptblock {
    $Response.Send("ab+cd")
}

New-PolarisRoute -Path "/ef*gh" -Method "GET" -Scriptblock {
    $Response.Send("ef*gh")
}

New-PolarisRoute -Path "/ij(kl)?m" -Method "GET" -Scriptblock {
    $Response.Send("ij(kl)?m")
}
```

You may now view them easily like this:

```shell
PS C:\my\Polaris> $Polaris = Get-Polaris
PS C:\my\Polaris> $Polaris.Routes

Path                         Regex                                      Method Scriptblock
----                         -----                                      ------ -----------
/users/:userId/books/:bookId ^/users/(?<userId>.+)/books/(?<bookId>.+)$ GET    ...
/userRegex/(?<userId>\d+)    /userRegex/(?<userId>\d+)                  GET    ...
/user/(?<userId>\d+)         ^/user/(?<userId>\d+)$                     GET    ...
/flights/:from-:to           ^/flights/(?<from>.+)\-(?<to>.+)$          GET    ...
/flights/:from.:to           ^/flights/(?<from>.+)\.(?<to>.+)$          GET    ...
/random.text                 ^/random\.text$                            GET    ...
/ij(kl)?m                    ^/ij(kl)?m$                                GET    ...
/about                       ^/about$                                   GET    ...
/12?34                       ^/12?34$                                   GET    ...
/ef*gh                       ^/ef.*gh$                                  GET    ...
/ab+cd                       ^/ab+cd$                                   GET    ...
/                            ^/$                                        GET    ...
```

Fixes #134
---

# Pull Request Creation Checklist

-   [x] Useful title (i.e. **Fix #XXX: Description** / **Feature #XXX: Description** / **RFC #XXX: Description**)
-   [x] Description of the changes you are making
-   [x] Added a link to the related Github issue in the description (i.e. type `Fixes #XXX` in the desciption of the PR)
-   [x] If you are still working on the code and would like some early feedback via the Pull Request process just add [WIP] to the beginning of your Pull Request title
-   [x] Checked the [Polaris GitHub workflow guidance](/GITHUB_GUIDANCE.md) to make sure my code has the latest changes made to the master branch
-   [x] Throw small party 🎉

### CODE REVIEW

**Comment by Tiberriver256** [comment]
Added some housekeeping. Killing some minor bugs.

🐛🐛🐜🐜
Fixes: #168 
Fixes: #171

File: Tests/e2e/PolarisServer.Tests.ps1 [hunk 1]
```diff
@@ -93,6 +98,12 @@ Describe "Test webserver use (E2E)" {
             $Result.StatusCode | Should Be 200
         }
 
+        It "test /empty route should return 204 empty" {
```
Comment (TylerLeonhardt) [inline Tests/e2e/PolarisServer.Tests.ps1]:
```suggestion
         It "test /empty route should return 204 No Content" {
```

File: lib/PolarisRoute.Class.ps1 [hunk 2]
```diff
@@ -0,0 +1,110 @@
+#
+# Copyright (c) Microsoft. All rights reserved.
+# Licensed under the MIT license. See LICENSE file in the project root for full license information.
+#
+
```
Comment (TylerLeonhardt) [inline lib/PolarisRoute.Class.ps1]: nit extra newline

File: lib/PolarisRoute.Class.ps1 [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    hidden static [string] SanitizePath([string]$Path) {
+        $SanitizedPath = $Path.TrimEnd('/')
+
+        if ([string]::IsNullOrEmpty($SanitizedPath)) { $SanitizedPath = "/" }
+
+        return $SanitizedPath
+    }
+
+    hidden static [string] SanitizePath([regex]$Path) {
```
Comment (TylerLeonhardt) [inline lib/PolarisRoute.Class.ps1]:
```suggestion
     hidden static [string] SanitizePath([RegEx]$Path) {
```

just make that's the consistent casing :)

File: lib/PolarisRoute.Class.ps1 [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+        $PathProperty = @{
+            Name        = "Path"
+            Value       = $Path_Get
+            SecondValue = $Path_Set
+            MemberType  = "ScriptProperty"
+        }
+
+        $this | Add-Member @PathProperty
```
Comment (TylerLeonhardt) [inline lib/PolarisRoute.Class.ps1]: Can you explain why the Add-Member stuff here is needed vs just having Path and Regex not be hidden?

File: Public/New-PolarisRoute.ps1 [hunk 5]
```diff
@@ -61,47 +86,46 @@ function New-PolarisRoute {
...
-    else {
-        CreateNewPolarisIfNeeded
-        if( -not $Polaris){
-            $Polaris = $Script:Polaris
-        }
 
-        if ( -not $Path.StartsWith( '/' ) ) {
-            $Path = '/' + $Path
-        }
+    if ( $Path.GetType().Name -eq "String" -and -not $Path.StartsWith( '/' ) ) {
```
[the remark this answers is not in the document]
Comment (Tiberriver256) [reply Public/New-PolarisRoute.ps1]: Fancy

File: Public/New-PolarisRoute.ps1 [hunk 6]
```diff
@@ -61,47 +86,46 @@ function New-PolarisRoute {
...
+    switch ( $PSCmdlet.ParameterSetName ) {
+        'Scriptblock' {
+            $Polaris.AddRoute( $Path, $Method, $Scriptblock )
+        }
+        'ScriptPath' {
+            $Script = Get-Content -Path $ScriptPath -Raw
+            $Polaris.AddRoute( $Path, $Method, [scriptblock]::Create($Script) )
         }
     }
+
```
[the remark this answers is not in the document]
Comment (Tiberriver256) [reply Public/New-PolarisRoute.ps1]: done

File: Public/New-PolarisRoute.ps1 [hunk 7]
```diff
@@ -32,12 +36,21 @@
     To view results, assuming default port:
     Start-Polaris
     Start-Process http://localhost:8080/helloworld
+.EXAMPLE
+    New-PolarisRoute -Path "helloworld" -Method "GET" -ScriptPath D:\Scripts\Example.ps1
+    To view results, assuming default port:
+    Start-Polaris
+    Start-Process http://localhost:8080/helloworld
+.EXAMPLE
```
[the remark this answers is not in the document]
Comment (Tiberriver256) [reply Public/New-PolarisRoute.ps1]: Indeed they were

File: lib/PolarisRoute.Class.ps1 (hunk 4 above)
Comment (Tiberriver256) [reply lib/PolarisRoute.Class.ps1]: Sure let me try and talk through my thought process here and then if it seems legit we'll add in some documentation into the code to clarify things a little bit.

On the current iteration of our build we only expose the "path" that was given to Polaris. The Regular Expression that the path is converted to at run-time isn't really exposed. If someone wants to figure out exactly how the path they fed Polaris is getting converted to RegEx they have to run this method against the path that they want.

https://github.com/PowerShell/Polaris/blob/a0117afb7308273d869167af2b3d8be4f5ed8fe4/lib/Polaris.Class.ps1#L211-L232

While that doesn't sound too bad it lends itself to a bit of mystery on how things are actually working under the hood.

What I wanted was for someone to be able to run the following commands:

```pwsh
> New-PolarisRoute -Path "/hello*/:name" -Method GET -Scriptblock { $Response.Send("Hello $($Parameters.Name)") }
> Get-PolarisRoute
```
and see the following output

```
Path          Regex                  Method Scriptblock
----          -----                  ------ -----------
/hello*/:name ^/hello.*/(?<name>.+)$ GET     $Response.Send("Hello $($Parameters.Name)")
```

Converting to RegEx when the path is set also should give a bit of a performance boost as the regular expression isn't being recalculated every time the path is run as we have here:

https://github.com/PowerShell/Polaris/blob/a0117afb7308273d869167af2b3d8be4f5ed8fe4/lib/Polaris.Class.ps1#L70

So essentially I had two requirements:

1. RegEx should be a read-only field
2. RegEx should be re-calculated internally every time the value of Path changes

If we were doing this in C# we could do both of those things easily on the class with custom getter and setters and an internal property. PowerShell classes however do not have getters and setters so what we are left with is the Add-Member method for achieving that. I found doing it in the constructor to be pretty effective per [this stackoverflow suggestion via Rob Traynere](https://stackoverflow.com/questions/40977472/is-it-possible-to-override-the-getter-setter-functions-in-a-powershell-5-class).

The results are as displayed above with the added bonus of being able to manipulate the paths directly if needed, and the RegEx is auto-updated. Also we provide a warning message if the user attempts to set the value of RegEx directly:

image

**Comment by Tiberriver256** [comment]
> If ($Path -is [string])...

How did I go all this time without learning of the `-is` operator!?! Love it, thanks @Wand3rvogel

**Comment by Tiberriver256** [comment]
After a 4 month long break I finally came back and got this done. Sorry for such a long delay @TylerLeonhardt. Let me know what you think of my comments above. I added in the documentation for routing so we should be ready to merge once we iron things out.

**Comment by Stephanevg** [comment]
Looking forward to this one :)

**Comment by TylerLeonhardt** [comment]
@Tiberriver256 can you resolve the conflicts? Also heads up - PS Summit is this week and I'll be very busy with that. I'll try to look at this PR again but no promises

**Comment by Tiberriver256** [comment]
@TylerLeonhardt - Merge conflicts resolved. I gotta get out to the summit one of these days meet a few of you famous people :)

**Comment by Tiberriver256** [comment]
@TylerLeonhardt - Gentle bump

**Comment by TylerLeonhardt** [comment]
Thanks for the nudge! Unfortunately my downtime didn't last long... now I'm on vacation/traveling to PSConfEU.

I've set a reminder to look when I get back

**Comment by Stephanevg** [comment]
It would be nice to have this one merged one day. Is this one still in the pipe @TylerLeonhardt ? ;)

File: lib/PolarisRoute.Class.ps1 (hunk 4 above)
Comment (TylerLeonhardt) [reply lib/PolarisRoute.Class.ps1]: Got around to reading this! I agree and think that's the right approach.

**Comment by TylerLeonhardt** [comment]
It is. Waiting on @Tiberriver256 to address the last comments. I can probably (finally) sit down and fix the release process (new internal requirements) once this goes in

**Comment by Tiberriver256** [comment]
Yep, thanks for the patience on this guys. Looking like I'll probably be able to get to it this weekend.

**Comment by Tiberriver256** [comment]
Should be all set @TylerLeonhardt

**Comment by TylerLeonhardt** [comment]
Looks good and CI is green :)

**Pull request merged**